### PR TITLE
added basic DLL support

### PR DESCRIPTION
**Pull request by VPaulV**

### CODE REVIEW

File: s2e_env/commands/new_project.py [hunk 1]
```diff
@@ -93,21 +95,24 @@ def handle(self, *args, **options):
             (Magic(magic_file=CGC_MAGIC), CGC_REGEX, CGCProjectConfiguration, 'i386'),
             (Magic(), ELF32_REGEX, LinuxProjectConfiguration, 'i386'),
             (Magic(), ELF64_REGEX, LinuxProjectConfiguration, 'x86_64'),
+            (Magic(), DLL32_REGEX, WindowsDllProjectConfiguration, 'i386'),
+            (Magic(), DLL64_REGEX, WindowsDllProjectConfiguration, 'x86_64'),
             (Magic(), PE32_REGEX, WindowsProjectConfiguration, 'i386'),
             (Magic(), PE64_REGEX, WindowsProjectConfiguration, 'x86_64'),
             (Magic(), MSDOS_REGEX, WindowsProjectConfiguration, 'i386')
-        ]
+       ]
```
Comment (adrianherrera) [inline s2e_env/commands/new_project.py]: Can you please fix the indentation back to how it was

File: s2e_env/commands/new_project.py [hunk 2]
```diff
@@ -93,21 +95,24 @@ def handle(self, *args, **options):
...
 
         # Check the target program against the valid file types
         for magic_check, regex, proj_class, arch in magic_checks:
             magic = magic_check.from_file(target_path)
             matches = regex.match(magic)
-
+
```
Comment (adrianherrera) [inline s2e_env/commands/new_project.py]: Undo this change to fix the indentation

File: s2e_env/commands/project_creation/__init__.py [hunk 3]
```diff
@@ -117,7 +117,6 @@ def handle(self, *args, **options):
             'use_symb_input_file': use_symb_input_file,
             'dynamically_linked': False,
             'modelled_functions': False,
-
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/__init__.py]: Undo change

File: s2e_env/commands/project_creation/__init__.py [hunk 4]
```diff
@@ -229,7 +228,7 @@ def _validate_and_create_project(self, options):
         templates = get_image_templates(img_build_dir)
 
         if not image:
-            image = self._guess_image(templates, options['target_arch'])
+            image = self._guess_image(templates, options['target_arch'], self._target_path)
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/__init__.py]: There is no point passing a `self` attribute to an instance method - you can access `self._target_path` directly in `_guest_image`.

File: s2e_env/commands/project_creation/__init__.py [hunk 5]
```diff
@@ -245,7 +244,7 @@ def _validate_and_create_project(self, options):
         # Create the project directory
         os.mkdir(self._project_dir)
 
-    def _guess_image(self, templates, target_arch):
+    def _guess_image(self, templates, target_arch, target_path):
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/__init__.py]: Remove `target_path` argument

File: s2e_env/commands/project_creation/__init__.py [hunk 6]
```diff
@@ -255,7 +254,7 @@ def _guess_image(self, templates, target_arch):
                     'a suitable image for a %s binary', target_arch)
 
         for k, v in templates.iteritems():
-            if self._configurator.is_valid_binary(target_arch, v['os']):
+            if self._configurator.is_valid_binary(target_arch, target_path, v['os']):
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/__init__.py]: `self._target_path`

File: s2e_env/commands/project_creation/config.py [hunk 7]
```diff
@@ -20,11 +20,10 @@
 SOFTWARE.
 """
 
-
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: Undo formatting

File: s2e_env/commands/project_creation/config.py [hunk 8]
```diff
@@ -53,12 +52,26 @@ def analyze(self, config):
...
 
 
+class WindowsDllProjectConfiguration(ProjectConfiguration):
+    BOOTSTRAP_TEMPLATE = 'bootstrap.windows_dll.sh'
+    LUA_TEMPLATE = 's2e-config.windows.lua'
+    PROJECT_TYPE = 'windows'
+
+    def is_valid_binary(self, target_arch, target_path, os_desc):
+        if not target_path.endswith('.dll'):
+            raise CommandError('Wrong DLL name. The .dll extension should be added')
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: "Invalid DLL name - requires .dll extension"

File: s2e_env/templates/bootstrap.windows_dll.sh [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    local SYMB_FILE
+
+    TARGET="$1"
+    SEED_FILE="$2"
+
+    ${S2EGET} "${SEED_FILE}"
+
+    SYMB_FILE="$(prepare_inputs ${SEED_FILE})"
+
+    ./${TARGET} {{ target_args | join(' ') }} > /dev/null 2> /dev/null
```
Comment (adrianherrera) [inline s2e_env/templates/bootstrap.windows_dll.sh]: This will not work for a DLL. Is this `execute_target_with_seed` function needed for DLLs?

File: s2e_env/templates/instructions.txt [hunk 10]
```diff
@@ -1,6 +1,21 @@
 Here are some hints to get started with your project:
 
 {#############################################################################}
+{% if dll_exports is defined %}
+
+The processed binary is a Windows DLL. The DLL will be loaded with default entry
```
Comment (adrianherrera) [inline s2e_env/templates/instructions.txt]: "with **the** default entry"

File: s2e_env/templates/instructions.txt [hunk 11]
```diff
@@ -1,6 +1,21 @@
...
+point set to DllEntryPoint.
+
+To change the default entry point please specify DLL_ENTRY_POINT in the
+bootstrap.sh
+
+The list of the detected exported functions is provided below. If name of the
+function is not avaiable its ordinal is printed:
+{{ dll_exports | join('\n') }}
```
Comment (adrianherrera) [inline s2e_env/templates/instructions.txt]: I wonder if instead of printing them all it would be better to just print the first 5 or so, then have `...` For some DLLs this list could be quite large

File: s2e_env/utils/pe.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+SOFTWARE.
+"""
+
+
+import pefile
+
+
+def get_dll_exports(target):
+    exports = []
+    print target
```
Comment (adrianherrera) [inline s2e_env/utils/pe.py]: why do you print this here? It should probably be removed

File: s2e_env/utils/pe.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    pe = pefile.PE(target)
+    dir(pe)
```
Comment (adrianherrera) [inline s2e_env/utils/pe.py]: What is the purpose of this?

File: s2e_env/commands/project_creation/config.py [hunk 14]
```diff
@@ -53,12 +52,26 @@ def analyze(self, config):
         pass
 
 
+class WindowsDllProjectConfiguration(ProjectConfiguration):
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: Can we please rename it to `WindowsDLLProjectConfiguration`. We use `CGC` (all upper-case), so I guess it makes sense for consistency

File: s2e_env/commands/project_creation/config.py [hunk 15]
```diff
@@ -20,11 +20,10 @@
 SOFTWARE.
 """
 
-
 import logging
 from s2e_env.command import CommandError
 from s2e_env.utils.elf import ELFAnalysis
-
+from s2e_env.utils.pe import get_dll_exports
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: I would do `from s2e_env.utils import pe` and then rename the `get_dll_exports` function to `get_exports`. Then when you use this function later on it is `pe.get_exports`, which I think is a bit cleaner

File: s2e_env/templates/bootstrap.windows_dll.sh (hunk 9 above)
Comment (VPaulV) [inline s2e_env/templates/bootstrap.windows_dll.sh]: no, had to remove

File: s2e_env/commands/project_creation/__init__.py [hunk 16]
```diff
@@ -324,8 +324,9 @@ def _symlink_target(self):
         logger.info('Creating a symlink to %s', self._target_path)
 
         target_file = os.path.basename(self._target_path)
-        os.symlink(self._target_path,
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/__init__.py]: I think that this change can be reverted to how it was previously

Comment (VPaulV) [inline s2e_env/commands/project_creation/__init__.py]: right, removed

File: s2e_env/commands/project_creation/config.py [hunk 17]
```diff
@@ -53,12 +54,26 @@ def analyze(self, config):
...
+    PROJECT_TYPE = 'windows'
+
+    def is_valid_binary(self, target_arch, target_path, os_desc):
+        if not target_path.endswith('.dll'):
+            raise CommandError('Invalid DLL name - requires .dll extension') 
+
+        return is_valid_arch(target_arch, os_desc) and 'pe' in os_desc['binary_formats']
+
+    def validate_configuration(self, config):
+        config['dll_exports'] = pe.get_exports(config['target_path'])
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: I think I made a comment last time, but is there any reason why this is in `validate_configuration`? Shouldn't it go in `analyze`?

Comment (VPaulV) [inline s2e_env/commands/project_creation/config.py]: You didn't. Sounds logical

Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: I did -> https://github.com/S2E/s2e-env/commit/5d4ac580256f035e05749eb3035640b55c8f4c94#commitcomment-22439460

Also, I this is where we should shorten the list of exports if it is too long - you don't want to display the entire list if it is very long to the user. Maybe something like:

```python
exports = pe.get_exports(config['target_path'])
if len(exports) <= 5:
    config['dll_exports'] = exports
else:
    config['dll_exports'] = exports[:5] + ['...']
```

This is one time where we do want to create a new list! :)

What do you think?

Comment (VPaulV) [inline s2e_env/commands/project_creation/config.py]: I think that there is no need to shorten it. Either have full list either don't have it at all

Comment (VPaulV) [inline s2e_env/commands/project_creation/config.py]: But it is just IMHO, you are the Boss =)

Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: okie-dokie

Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: Let's leave it as-is for now.

File: s2e_env/commands/project_creation/config.py [hunk 18]
```diff
@@ -53,12 +54,26 @@ def analyze(self, config):
...
 
 
+class WindowsDLLProjectConfiguration(ProjectConfiguration):
+    BOOTSTRAP_TEMPLATE = 'bootstrap.windows_dll.sh'
+    LUA_TEMPLATE = 's2e-config.windows.lua'
+    PROJECT_TYPE = 'windows'
+
+    def is_valid_binary(self, target_arch, target_path, os_desc):
+        if not target_path.endswith('.dll'):
+            raise CommandError('Invalid DLL name - requires .dll extension')
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: Remove the trailing whitespace at the end of this line

File: s2e_env/commands/new_project.py [hunk 19]
```diff
@@ -31,7 +31,8 @@
 from s2e_env.manage import call_command
 from s2e_env.commands.project_creation import BaseProject
 from s2e_env.commands.project_creation.config import \
-    CGCProjectConfiguration, LinuxProjectConfiguration, WindowsProjectConfiguration
+    CGCProjectConfiguration, LinuxProjectConfiguration, WindowsProjectConfiguration, \
+    WindowsDLLProjectConfiguration
```
Comment (adrianherrera) [inline s2e_env/commands/new_project.py]: There is no need for this newline here

File: s2e_env/templates/instructions.txt [hunk 20]
```diff
@@ -1,6 +1,21 @@
...
+
+The processed binary is a Windows DLL. The DLL will be loaded with the default
+entry point set to DllEntryPoint.
+
+To change the default entry point please specify DLL_ENTRY_POINT in the
+bootstrap.sh
+
+The list of the detected exported functions is provided below. If name of the
+function is not avaiable its ordinal is printed:
+{{ dll_exports | join(', ') }}
```
Comment (adrianherrera) [inline s2e_env/templates/instructions.txt]: Do you think you should join with a comma or a newline? I just did some testing on kernel32.dll and it is very long and difficult to read.

File: s2e_env/templates/instructions.txt [hunk 21]
```diff
@@ -1,6 +1,21 @@
...
 {#############################################################################}
+{% if dll_exports is defined %}
+
+The processed binary is a Windows DLL. The DLL will be loaded with the default
+entry point set to DllEntryPoint.
+
+To change the default entry point please specify DLL_ENTRY_POINT in the
+bootstrap.sh
+
+The list of the detected exported functions is provided below. If name of the
```
Comment (adrianherrera) [inline s2e_env/templates/instructions.txt]: "**A** list of exported functions is provided below. If **the** name of the"

File: s2e_env/commands/project_creation/config.py [hunk 22]
```diff
@@ -53,12 +54,26 @@ def analyze(self, config):
...
+
+    def is_valid_binary(self, target_arch, target_path, os_desc):
+        if not target_path.endswith('.dll'):
+            raise CommandError('Invalid DLL name - requires .dll extension') 
+
+        return is_valid_arch(target_arch, os_desc) and 'pe' in os_desc['binary_formats']
+
+    def analyze(self, config):
+        config['dll_exports'] = pe.get_exports(config['target_path'])
+
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: Can you add an extra newline here please

File: s2e_env/commands/project_creation/config.py [hunk 23]
```diff
@@ -53,12 +54,26 @@ def analyze(self, config):
         pass
 
 
+class WindowsDLLProjectConfiguration(ProjectConfiguration):
```
Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: Can you also implement `validate_configuration` for this class too. If `use_seeds` has been set to `True`, it should display a warning (via `logger.warn`) that says something along the lines of "Seeds have been enabled, however they are not supported for DLLs. This flag will be ignored".

Then set `config['use_seeds'] = False`.

File: s2e_env/commands/new_project.py [hunk 24]
```diff
@@ -44,7 +45,8 @@
 PE32_REGEX = re.compile(r'^PE32 executable')
 PE64_REGEX = re.compile(r'^PE32\+ executable')
 MSDOS_REGEX = re.compile(r'^MS-DOS executable')
-
+DLL32_REGEX = re.compile(r'^PE32 executable \(DLL\)')
+DLL64_REGEX = re.compile(r'^PE32\+ executable \(DLL\)')
```
Comment (adrianherrera) [inline s2e_env/commands/new_project.py]: Can you add an extra newline here please

**Comment by adrianherrera** [comment]
Paul never finished the PR, so I've fixed it up. It should be ready to merge now.

File: s2e_env/commands/project_creation/config.py (hunk 23 above)
Comment (adrianherrera) [inline s2e_env/commands/project_creation/config.py]: Done.

**Comment by VPaulV** [comment]
Damn! Sorry I thought I have pushed changes :/

**Comment by Epeius** [comment]
I will review this.
